### PR TITLE
Backend update progress api

### PR DESCRIPTION
**Pull request by iamAbasu7**
tried to fix the progress api issue #46

### CODE REVIEW

File: packages/backend/package.json [hunk 1]
```diff
@@ -24,6 +24,7 @@
     "@types/express": "^4.17.17",
     "@types/http-errors": "^2.0.1",
     "@types/node": "^18.14.0",
+    "@types/node-cron": "^3.0.7",
```
Comment (jastisriradheshyam) [inline packages/backend/package.json:27]: remove this unused dependency

File: packages/backend/package.json [hunk 2]
```diff
@@ -41,6 +42,7 @@
     "http-errors": "^2.0.0",
     "joi": "^17.8.1",
     "knex": "^2.4.2",
+    "node-cron": "^3.0.2",
```
Comment (jastisriradheshyam) [inline packages/backend/package.json:45]: remove this unused dependency

File: packages/backend/src/modules/network.ts [hunk 3]
```diff
@@ -1,4 +1,6 @@
-import { Network, NetworkInfo } from './models/network.js';
+/* eslint-disable no-await-in-loop */
+// import cron from 'node-cron';
```
Comment (jastisriradheshyam) [inline packages/backend/src/modules/network.ts:2]: remove this unused dependency

File: packages/backend/src/modules/network.ts [hunk 4]
```diff
@@ -15,11 +17,42 @@ import { removeInProgressNetwork } from './exec_run.js';
...
+    // console.log(sortedResult[i].name);
+    const runInfo = new ExecRun();
+    const result = await runInfo.getStatusCode(sortedResult[i].name);
+    // console.log(result);
+    for (let j = 0; j < result.length; j++) {
+      const status = result[j].statusCode;
+      let state: NetworkState = 'failed';
+      if (status === null) {
+        state = 'creating';
+        await network.updateNetworkStatus(sortedResult[i].name, state);
```
Comment (jastisriradheshyam) [inline packages/backend/src/modules/network.ts:33]: get final state value and call this once, also remove await as this process can be done asynchronously

File: packages/backend/src/modules/network.ts [hunk 5]
```diff
@@ -15,11 +17,42 @@ import { removeInProgressNetwork } from './exec_run.js';
...
 const getNetworkPath = (networkName: string): string => `${ZOMBIENET_NETWORKS_COLLECTION_DIR}/${networkName}`;
 
-export const showNetworkProgress = async (
+async function networkStatusUpdate() {
+  const network = new Network();
+  const sortedResult = await network.getAll();
+  for (let i = 0; i < sortedResult.length; i++) {
+    // console.log(sortedResult[i].name);
+    const runInfo = new ExecRun();
+    const result = await runInfo.getStatusCode(sortedResult[i].name);
```
Comment (jastisriradheshyam) [inline packages/backend/src/modules/network.ts:26]: do not use ExecRun class as it is not applicable to single ExecRun record, and also just get the latest one (by created at field)

File: packages/backend/src/modules/network.ts [hunk 6]
```diff
@@ -15,11 +17,42 @@ import { removeInProgressNetwork } from './exec_run.js';
 
 const getNetworkPath = (networkName: string): string => `${ZOMBIENET_NETWORKS_COLLECTION_DIR}/${networkName}`;
 
-export const showNetworkProgress = async (
+async function networkStatusUpdate() {
+  const network = new Network();
+  const sortedResult = await network.getAll();
```
Comment (jastisriradheshyam) [inline packages/backend/src/modules/network.ts:22]: do not use Network class as it is not applicable to single network record, create a separate function for get all the network records.

File: packages/backend/src/modules/models/network.ts [hunk 7]
```diff
@@ -85,11 +103,19 @@ export class Network {
...
-  async findNetworkProgress(): Promise<any> {
+  // async findNetworkProgress(): Promise<any> {
+  //   // console.log(this.id);
+  //   const [result] = await this.db()
+  //     .select('network_state')
+  //     .where('name', this.name);
+  //   return result;
+  // }
+
+  async getNetworkState(network_name: string): Promise<any> {
```
Comment (jastisriradheshyam) [inline packages/backend/src/modules/models/network.ts:114]: only return network state value not the object containing network state

File: packages/backend/src/api/larch/network/controllers.ts [hunk 8]
```diff
@@ -148,9 +148,8 @@ export const networkTestController = async (req: Request, res: Response) => {
 };
 
 export const progressController = async (req: Request, res: Response) => {
-  const searchNetwork: string | any = req.query.networkName;
-  const updatedNetworkName = searchNetwork.replace(/\s/g, '');
-  const result = await showNetworkProgress(updatedNetworkName);
+  const networkName = typeof req.query.networkName === 'string' ? req.query.networkName : '';
+  const result = await progressNetwork(networkName);
```
Comment (jastisriradheshyam) [inline packages/backend/src/api/larch/network/controllers.ts:152]: directly use Network object and use getNetworkState function don't have to use an abstraction function for small functionality and also check if the network exists.

File: packages/backend/src/modules/models/network.ts [hunk 9]
```diff
@@ -67,6 +67,24 @@ export class Network {
     };
   }
 
+  async getAll(): Promise<any> {
```
Comment (jastisriradheshyam) [inline packages/backend/src/modules/models/network.ts:70]: remove it from here and create a separate function.

File: packages/backend/src/modules/models/network.ts [hunk 10]
```diff
@@ -67,6 +67,24 @@ export class Network {
...
+
+  // async getNetworkId(): Promise<string> {
+  //   const [result] = await this.db()
+  //     .select('*')
+  //     .where('name', this.name);
+  //   return {
+  //     id: this.name,
+  //     ...result,
+  //   };
+  // }
```
Comment (jastisriradheshyam) [inline packages/backend/src/modules/models/network.ts:86]: remove commented code

File: packages/backend/src/modules/models/network.ts [hunk 11]
```diff
@@ -85,11 +103,19 @@ export class Network {
...
   }
 
-  async findNetworkProgress(): Promise<any> {
+  // async findNetworkProgress(): Promise<any> {
+  //   // console.log(this.id);
+  //   const [result] = await this.db()
+  //     .select('network_state')
+  //     .where('name', this.name);
+  //   return result;
+  // }
```
Comment (jastisriradheshyam) [inline packages/backend/src/modules/models/network.ts:112]: remove commented code

File: packages/backend/src/modules/models/network.ts [hunk 12]
```diff
@@ -85,11 +103,19 @@ export class Network {
...
+
+  async getNetworkState(network_name: string): Promise<any> {
     // console.log(this.id);
     const [result] = await this.db()
       .select('network_state')
-      .where('name', this.name);
+      .where('name', network_name);
```
Comment (jastisriradheshyam) [inline packages/backend/src/modules/models/network.ts:118]: get network name from `this`

File: packages/backend/src/modules/models/network.ts [hunk 13]
```diff
@@ -108,6 +134,17 @@ export class Network {
         network_state: state,
       });
   }
+
+  async updateNetworkStatus(
+    networkName: string,
```
Comment (jastisriradheshyam) [inline packages/backend/src/modules/models/network.ts:139]: get network name from `this`, don't get the network name separately, remove this parameter.

File: packages/backend/src/modules/models/network.ts [hunk 14]
```diff
@@ -108,6 +134,17 @@ export class Network {
...
+    state: NetworkState,
+  ): Promise<void> {
+    await this.db()
+      .where('name', networkName)
```
Comment (jastisriradheshyam) [inline packages/backend/src/modules/models/network.ts:143]: get network name from `this`

File: packages/backend/src/modules/models/user_operation.ts [hunk 15]
```diff
@@ -62,6 +62,13 @@ export class UserOperation {
       .delete()
       .where('id', this.id);
   }
+
+  async sortByCreatedAt(): Promise<void> {
+    const result = await this.db()
+      .select('*')
+      .orderBy('created_at', 'desc');
+    console.log(result);
+  }
```
Comment (jastisriradheshyam) [inline packages/backend/src/modules/models/user_operation.ts:71]: create a separate function, don't include in Network class

File: packages/backend/src/modules/network.ts [hunk 16]
```diff
@@ -15,11 +17,42 @@ import { removeInProgressNetwork } from './exec_run.js';
...
+export const progressNetwork = async (
   networkName: string,
-): Promise<void> => {
-  const network = new Network(networkName);
-  const result = await network.findNetworkProgress();
+): Promise<any> => {
+  const network = new Network();
+  const result = await network.getNetworkState(networkName);
   return result;
 };
```
Comment (jastisriradheshyam) [inline packages/backend/src/modules/network.ts:57]: remove this function since, Network object will be use in controller

**Review by jastisriradheshyam — CHANGES_REQUESTED** [review]

**Review by radhe-zeeve — APPROVED** [review]

**Review by jastisriradheshyam — APPROVED** [review]

**Pull request merged**